src/controllers/krate/owners.rs

```diff
@@ -171,11 +172,44 @@ async fn modify_owners(
                     if owners.iter().any(login_test) {
                         return Err(bad_request(format_args!("`{login}` is already an owner")));
                     }
-                    let (msg, maybe_email) = krate.owner_add(&app, conn, user, login)?;
-                    msgs.push(msg);
 
-                    if let Some(v) = maybe_email {
-                        emails.push(v);
+                    match krate.owner_add(&app, conn, user, login) {
+                        // A user was successfully invited, and they must accept
+                        // the invite, and a best-effort attempt should be made
+                        // to email them the invite token for one-click
+                        // acceptance.
+                        Ok(NewOwnerInvite::User(user, token)) => {
+                            msgs.push(format!(
+                                "user {} has been invited to be an owner of crate {}",
+                                user.gh_login, krate.name,
+                            ));
+
+                            if let Some(recipient) = user.verified_email(conn).ok().flatten() {
+                                emails.push(OwnerInviteEmail {
+                                    recipient_email_address: recipient,
+                                    user_name: user.gh_login.clone(),
+                                    domain: app.emails.domain.clone(),
+                                    crate_name: krate.name.clone(),
+                                    token,
+                                });
+                            }
+                        }
+
+                        // A team was successfully invited. They are immediately
+                        // added, and do not have an invite token.
+                        Ok(NewOwnerInvite::Team(team)) => msgs.push(format!(
+                            "team {} has been added as an owner of crate {}",
+                            team.login, krate.name
+                        )),
+
+                        // This user has a pending invite.
+                        Err(OwnerAddError::AlreadyInvited(user)) => msgs.push(format!(
+                            "user {} already has a pending invitation to be an owner of crate {}",
+                            user.gh_login, krate.name
+                        )),
+
+                        // An opaque error occurred.
+                        Err(OwnerAddError::AppError(e)) => return Err(e),
```

Comment:
The controller can now inspect the model call result, and generate the appropriate API response / emails.